### PR TITLE
DAOS-5865 control: fix printing of ssd fw revision

### PR DESCRIPTION
**Pull request by tanabarr**
go-spdk bindings were printing unwanted tailing characters after
firmware revision and disrupting formatting of table in dmg command
output.

- convert returned controller details to ASCII to follow NVMe spec
- add src/common/control.c to implement src/include/daos_srv/control.h
- avoid duplicate controller entries in storage scan --nvme-health
- provide pci include list on bdev scan when iosrv online

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Comment by tanabarr** [comment]
subsequent output with this PR:
```
tanabarr@wolf-151:~/projects/daos_m> install/bin/dmg storage scan --verbose -i
---------
localhost
---------
SCM Namespace Socket ID Capacity
------------- --------- --------
pmem0         0         4.3 TB
pmem1         1         4.3 TB

NVMe PCI     Model                      FW Revision Socket ID Capacity
--------     -----                      ----------- --------- --------
0000:65:00.0 SAMSUNG MZWLJ3T8HBLS-00007 EPK98B5Q    0         3.8 TB
0000:67:00.0 SAMSUNG MZWLJ3T8HBLS-00007 EPK98B5Q    0         3.8 TB
0000:69:00.0 SAMSUNG MZWLJ3T8HBLS-00007 EPK98B5Q    0         3.8 TB
0000:6b:00.0 SAMSUNG MZWLJ3T8HBLS-00007 EPK98B5Q    0         3.8 TB
0000:e5:00.0 SAMSUNG MZWLJ3T8HBLS-00007 EPK98B5Q    1         3.8 TB
0000:e9:00.0 SAMSUNG MZWLJ3T8HBLS-00007 EPK98B5Q    1         3.8 TB
```

File: src/control/lib/spdk/src/nvme_control_common.c [hunk 1]
```diff
@@ -248,28 +248,48 @@ _discover(prober probe, bool detach, health_getter get_health)
...
-			   cdata->mn);
-	if (written >= sizeof(cdst->model))
+	if (buf_sz >= BUFLEN)
 		return -NVMEC_ERR_CHK_SIZE;
 
-	written = snprintf(cdst->serial, sizeof(cdst->serial), "%-20.20s",
-			   cdata->sn);
-	if (written >= sizeof(cdst->serial))
-		return -NVMEC_ERR_CHK_SIZE;
+	/* Trim trailing spaces */
```
Comment (NiuYawei) [inline src/control/lib/spdk/src/nvme_control_common.c:261]: isspace() could be used if there are other trailing characters like '\t' or '\n' to be trimmed.

File: src/control/lib/spdk/src/nvme_control_common.c [hunk 2]
```diff
@@ -248,28 +248,48 @@ _discover(prober probe, bool detach, health_getter get_health)
...
+	while (buf_sz > 0 && str[buf_sz - 1] == ' ')
+		buf_sz--;
 
-	written = snprintf(cdst->fw_rev, sizeof(cdst->fw_rev), "%s", cdata->fr);
-	if (written >= sizeof(cdst->fw_rev))
-		return -NVMEC_ERR_CHK_SIZE;
+	while (buf_sz--) {
+		if (*str >= 0x20 && *str <= 0x7E)
```
Comment (NiuYawei) [inline src/control/lib/spdk/src/nvme_control_common.c:266]: isprint() could be used if you want check if it's printable characters.

File: src/control/lib/spdk/src/nvme_control_common.c [hunk 3]
```diff
@@ -248,28 +248,48 @@ _discover(prober probe, bool detach, health_getter get_health)
...
+			dst[i] = (char)*str;
+		else
+			dst[i] = '.';
+		str++, i++;
+	}
+	dst[i] = '\0';
```
Comment (NiuYawei) [inline src/control/lib/spdk/src/nvme_control_common.c:272]: I think we have to be very careful here: what's the size of 'dst'? If the 'dst' size is equal to size of 'buf', it'll overflow when nothing is trimmed in 'buf'.

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 1 above)
Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c:261]: I'm following the recommendations made in SPDK identity example here as it will be hopefully aligned with what is expected in the NVMe spec.

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 2 above)
Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c:266]: as above

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 3 above)
Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c:272]: dst size will always be BUFLEN and buf_sz < BUFLEN is checked, but I will change to pass in dst_sz to be more explicit

File: src/control/lib/spdk/src/nvme_control_common.c [hunk 4]
```diff
@@ -248,28 +248,48 @@ _discover(prober probe, bool detach, health_getter get_health)
...
+copy_ascii(char *dst, const void *buf, size_t buf_sz)
 {
-	int	written;
+	const uint8_t	*str = buf;
+	int		 i = 0;
 
-	written = snprintf(cdst->model, sizeof(cdst->model), "%-20.20s",
-			   cdata->mn);
-	if (written >= sizeof(cdst->model))
+	if (buf_sz >= BUFLEN)
```
Comment (kjacque) [inline src/control/lib/spdk/src/nvme_control_common.c:258]: This strikes me as a bit odd. I'd expect equality to BUFLEN to be okay. I noticed  in the caller that we're getting the size of the incoming buffers with sizeof, which is right to me. But in that case, even if the dest size == BUFLEN, src size == BUFLEN could be fine, as the actual string we come up with could be shorter than the buffer length.

I had to read on to realize you were using buf_sz to mean the length of the string inside the buf. One suggestion, for clarity in differentiating the two: add a new variable called len, that starts at buf_sz - 1, and use len to represent the string length.

File: src/control/lib/spdk/src/nvme_control_common.c [hunk 5]
```diff
@@ -248,28 +248,48 @@ _discover(prober probe, bool detach, health_getter get_health)
...
-	if (written >= sizeof(cdst->serial))
-		return -NVMEC_ERR_CHK_SIZE;
+	/* Trim trailing spaces */
+	while (buf_sz > 0 && str[buf_sz - 1] == ' ')
+		buf_sz--;
 
-	written = snprintf(cdst->fw_rev, sizeof(cdst->fw_rev), "%s", cdata->fr);
-	if (written >= sizeof(cdst->fw_rev))
-		return -NVMEC_ERR_CHK_SIZE;
+	while (buf_sz--) {
```
Comment (kjacque) [inline src/control/lib/spdk/src/nvme_control_common.c:265]: Given the logic inside the loop, this should probably be a for-loop with the length of the string as the max.

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 1 above)
Comment (kjacque) [reply src/control/lib/spdk/src/nvme_control_common.c:261]: I don't see any harm in using isspace() in this case. We probably don't want to print any of those other whitespace characters either, if they did show up here.

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 3 above)
Comment (kjacque) [reply src/control/lib/spdk/src/nvme_control_common.c:272]: IMO we should be checking the size after we've trimmed the incoming buf, not before. Then we'll know for sure whether the string will fit.

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 1 above)
Comment (mjmac) [reply src/control/lib/spdk/src/nvme_control_common.c:261]: Out of curiosity, I went and looked at the NVMe spec, where it defines the convention for string parameters:

> 
> Some parameters are defined as an ASCII string. ASCII strings shall contain only code values 20h through
> 7Eh. For the string “Copyright”, the character “C” is the first byte, the character “o” is the second byte, etc.
> The string is left justified and shall be padded with spaces (ASCII character 20h) to the right if necessary.
> A hexadecimal ASCII string is an ASCII string that uses a subset of the code values: “0” to “9”, “A” to “F”
> uppercase, and “a” to “f” lowercase.
Given that, it seems like you could just stop copying as soon as you hit the first space character.

Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c]: as above, would like to stick with the methodology specified in the SPDK example code unless we have a really good reason to switch from it given that is the framework we are receiving information from

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 3 above)
Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c]: addressed

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 4 above)
Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c]: addressed

File: src/control/lib/spdk/src/nvme_control_common.c (hunk 5 above)
Comment (tanabarr) [reply src/control/lib/spdk/src/nvme_control_common.c]: addressed

**Review by NiuYawei — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

File: src/control/server/storage/bdev/provider.go [hunk 6]
```diff
@@ -177,30 +178,44 @@ func (resp *ScanResponse) filter(pciFilter ...string) *ScanResponse {
...
+				return
+			}
+
+			msg := fmt.Sprintf("%s bdev scan cache (%d",
+				action, len(resp.Controllers))
+
 			if resp == nil {
+				p.log.Debugf("%s devices)", msg)
 				return
 			}
```
Comment (mjmac) [inline src/control/server/storage/bdev/provider.go:197]: This fmt.Sprintf() will segfault if resp is nil. Probably just need to put the test for nil before it.

**Review by mjmac — CHANGES_REQUESTED** [review]
One issue that needs to be addressed. Looks fine otherwise.

**Review by kjacque — COMMENTED** [review]
LGTM aside from Mike's comment.

**Comment by tanabarr** [comment]
One or more failures in all of these suites:
DaosCoreTest
DaosCoreTestRebuild
DAOS Array API tests
DAOS KV API tests
DAOS rebuild tests

Related known CI failures:
DAOS-5882
DAOS-5884
DAOS-5854
DAOS-5852

File: src/control/server/storage/bdev/provider.go (hunk 6 above)
Comment (tanabarr) [reply src/control/server/storage/bdev/provider.go:197]: Thanks good catch (unlikely to occur in practice as we previously check for error but worth fixing anyway).

Taking the opportunity to merge in following PR https://github.com/daos-stack/daos/pull/3712 which moves inline `parse_ascii()` definition to `src/common/control.c` (where we can put any function definitions that need to be shared between control and data planes), this was requested by @NiuYawei . Hoping @jolivier23 can verify the `scons` changes look okay. 

#3712 also provides PCI address include list to SPDK on `scan --nvme-health` which avoids scan iteration for one instance trying to access devices claimed by the other running instance, fixes cosmetic issue which displays an "already claimed" error.

**Comment by tanabarr** [comment]
gatekeeper please use PR description and title when landing, thanks

**Review by kjacque — APPROVED** [review]

**Review by mjmac — APPROVED** [review]
LGTM

File: src/bio/SConscript [hunk 7]
```diff
@@ -28,8 +28,9 @@ def scons():
     # Other libs
     libs += ['numa', 'dl', 'smd']
 
-    bio = daos_build.library(denv, "bio", Glob('*.c'), install_off="../..",
-                             LIBS=libs)
+    tgts = Glob('*.c')
```
Comment (jolivier23) [inline src/bio/SConscript:31]: I know it's not your fault but I really don't like Glob.

File: src/control/lib/spdk/SConscript [hunk 8]
```diff
@@ -28,7 +28,9 @@ def scons():
 
     senv.nc = senv.Object("src/nvme_control.c")
     senv.ncc = senv.Object("src/nvme_control_common.c")
-    denv.nvmecontrol = senv.StaticLibrary("nvme_control", [senv.nc, senv.ncc],
+    senv.cc = senv.Object("../../../common/control.c")
```
Comment (jolivier23) [inline src/control/lib/spdk/SConscript:31]: I have a minor preference for exporting the file

something like control_tgts = File("control.c")
Export ('control_tgts')

then 
Import('control_tgts') # note that tgts suffix will avoid any issues with pylint checking.
and pass that to your compile line

**Review by jolivier23 — APPROVED** [review]

File: src/control/lib/spdk/SConscript (hunk 8 above)
Comment (tanabarr) [reply src/control/lib/spdk/SConscript]: done

File: src/bio/SConscript [hunk 9]
```diff
@@ -28,8 +28,9 @@ def scons():
...
+    tgts += ['../common/control.c']
```
Comment (jolivier23) [inline src/bio/SConscript:32]: control_tgts

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/bio/SConscript (hunk 9 above)
Comment (tanabarr) [reply src/bio/SConscript]: thanks, done

**Review by jolivier23 — APPROVED** [review]
scons changes LGTM

**Review by kjacque — APPROVED** [review]

**Comment by tanabarr** [comment]
CI failures :
3 failures that look unrelated to this PR
using filter "labels = intermittent_test_failure AND status != Resolved" can't seem to find a match
all failures related to "failed to destroy container"

**Comment by mjmac** [comment]
There's actually only one failure, seems to be [DAOS-5851](https://jira.hpdd.intel.com/browse/DAOS-5851). The failed test accounting is sometimes kind of weird and counts the parents of a failed test as extra failures.
